modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,337 @@
+    // If found, add the node to boundary_orderred_node_list
+    if (result != boundary_node_assm.end())
+    {
+      boundary_orderred_node_list.push_back(match_first ? (*result).second : (*result).first);
+      boundary_node_assm.erase(result);
+    }
+    // If there are still elements in boundary_node_assm and result ==
+    // boundary_node_assm.end(), this means the boundary is not a loop, the
+    // boundary_orderred_node_list is flipped and try the other direction that has not
+    // been examined yet.
+    else
+    {
+      if (isFlipped)
+      {
+        // Flipped twice; this means the boundary has at least two segments.
+        return false;
+      }
+      // mark the first flip event.
+      isFlipped = true;
+      std::reverse(boundary_orderred_node_list.begin(), boundary_orderred_node_list.end());
+      // As this iteration is wasted, set the iterator backward
+      i--;
+    }
+  }
+  // If the code ever gets here, boundary_node_assm is empty.
+  // If the isFlipped == true, the boundary is not a loop.
+  // This is not done inside the loop just for some potential applications in the future.
+  if (isFlipped)
+    return false;
+  // It the boundary is a loop, check if azimuthal angles change monotonically
+  else
+  {
+    // Utilize cross product here.
+    // If azimuthal angles change monotonically,
+    // the z components of the cross products are always negative or positive.
+    std::vector<Real> orderred_node_azi_list;
+    for (unsigned int i = 0; i < boundary_orderred_node_list.size() - 1; i++)
+    {
+      orderred_node_azi_list.push_back(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt)
+              .cross(*mesh.node_ptr(boundary_orderred_node_list[i + 1]) - origin_pt)(2));
+      // Use this opportunity to calculate maximum radius
+      max_node_radius = std::max(
+          (*mesh.node_ptr(boundary_orderred_node_list[i]) - origin_pt).norm(), max_node_radius);
+    }
+    std::sort(orderred_node_azi_list.begin(), orderred_node_azi_list.end());
+    if (orderred_node_azi_list.front() * orderred_node_azi_list.back() < 0.0)
+      return false;
+    else
+      return true;
+  }
+}
+
+bool
+PeripheralRingMeshGenerator::isBoundaryValid(ReplicatedMesh & mesh,
+                                             const Point origin_pt,
+                                             const boundary_id_type bid) const
+{
+  Real dummy_max_node_radius;
+  return isBoundaryValid(mesh, dummy_max_node_radius, origin_pt, bid);
+}
+
+bool
+PeripheralRingMeshGenerator::isExternalBoundary(ReplicatedMesh & mesh,
+                                                const boundary_id_type bid) const
+{
+  BoundaryInfo & boundary_info = mesh.get_boundary_info();
+  auto side_list = boundary_info.build_side_list();
+  for (unsigned int i = 0; i < side_list.size(); i++)
+  {
+    if (std::get<2>(side_list[i]) == bid)
+      if (mesh.elem_ptr(std::get<0>(side_list[i]))->neighbor_ptr(std::get<1>(side_list[i])) !=

```

Comment:
we ll have to make sure the mesh is prepared for use or at least that find_neighbors has been called